File: bob.rb
class Bob
  def hey(noise)
    noise = Noise.new(noise)
    if noise.silence?
      "Fine. Be that way!"
    elsif noise.shouting?
      'Woah, chill out!'
    elsif noise.question?
      'Sure.'
    else
      "Whatever."
    end
  end
end

class Noise
  attr_reader :noise
  def initialize(noise)
    @noise = noise
  end

  def shouting?
    noise.upcase == noise
  end

  def question?
    noise =~ /\?$/
  end

  def silence?
    (noise =~ /^\s*$/ || noise.nil?)
  end
end

File: javascript/bob/bob.js
var Bob = function(){
    this.hey = function(noise){
        var noise = new Noise(noise);

        if (noise.silence()){
            return "Fine. Be that way!";
        }

        if (noise.shouting()){
            return "Woah, chill out!";
        }

        if (noise.question()) {
            return "Sure.";
        }

        return "Whatever.";
    }
};

var Noise = function(noise){
    var noise = noise;

    this.shouting = function(){
        return (noise.toUpperCase() == noise);
    }

    this.question = function(){
        return (noise.slice(-1) == "?");
    }

    this.silence = function(){
        return (noise.trim() == false);
    }
}

module.exports = Bob;
